File: 2023/12.hs
{-# OPTIONS_GHC -Wall #-}
module Main
    where

import MyAOCLib

parse :: [String] -> [(String, [Int])]
parse = map parseLine
    where
        parseLine :: String -> (String, [Int])
        parseLine line = (symbols, numbers)
            where
                [symbols, rawNumbers] = split ' ' line
                numbers = map read $ split ',' rawNumbers

countSpringConfigurations :: (String, [Int]) -> Int
countSpringConfigurations ("", l)
    | null l    = 1
    | otherwise = 0
countSpringConfigurations (xs, [])
    | '#' `elem` xs = 0
    | otherwise     = 1
-- Case: Flere instruktioner. Needs a at least instruction+2 space
countSpringConfigurations (xs, instruction:instruction2:instructions)
    | length xs < instruction + 2   = 0
countSpringConfigurations (xs, [instruction])
    | instruction > length xs   = 0
    | (instruction == length xs) && (all (`elem` "#?") xs)  = 1
    | (instruction == length xs) = 0
-- Case 3 here is if xs contains some '.', but the instruction
-- needs them all to be # or ?.

{- To take 'instruction' many #'s, we need
 - exactly (instruction-1) #'s in xs and that
 - xs!!(instruction-1) is '.'. In that case we
 - eat x, (instruction-1) #'s and a '.'
 -}

countSpringConfigurations (x:xs, instruction:instructions)
    | x == '.'      = countSpringConfigurations (xs, (instruction:instructions))
    | (x == '#')    = if (all (`elem` "#?") $ take (instruction-1) xs) && (xs!!(instruction-1) `elem` ".?") then
                        countSpringConfigurations (drop instruction xs, instructions)
                        else 0
    | x == '?'      = (countSpringConfigurations ('.':xs, instruction:instructions))
                        + (countSpringConfigurations ('#':xs, instruction:instructions))

part1 l = sum $ map countSpringConfigurations l

multiplyInput :: [(String, [Int])] -> [(String, [Int])]
multiplyInput lines = map multiplyInputLine lines
    where
        multiplyInputLine (xs, instructions) = (repeatXs, repeatInstructions)
            where
                repeatXs = tail $ foldl (++) "" (take 5 $ repeat ('?':xs))
                repeatInstructions = foldl (++) [] (take 5 $ repeat instructions)
                

part2 = part1 . multiplyInput

main = do
    rawSampleInput <- (readLines "12sample.in")
    rawInput <- (readLines "12.in")

    putStrLn "Hello, World!"
    putStrLn $ show $ part1 $ parse $ rawSampleInput
    putStrLn $ show $ part1 $ parse $ rawInput
    putStrLn "Part 2"
    putStrLn $ show $ part2 $ parse $ rawSampleInput
    putStrLn $ show $ part2 $ parse $ rawInput

File: 2023/12.py
import sys
import functools

@functools.lru_cache
def countConfigs(xs, instructions):
    if len(xs) < sum(instructions) + len(instructions) - 1:
        return 0
    elif xs == "":
        if instructions == ():
            return 1
        return 0
    elif instructions == ():
        if "#" in xs:
            return 0
        return 1
    elif len(instructions) == 1:
        if (
                (instructions[0] == len(xs))
                and
                all(map(lambda x: x in "#?", xs))
        ):
            return 1
#        if (
#                all(map(lambda x: x in "#?", xs[:instructions[0]]))
#                and
#                all(map(lambda x: x in "#?", xs[:instructions[0]]))
#            ):
#            return 1
#        return 0
    if xs[0] == '.':
        return countConfigs(xs[1:], instructions)
    elif xs[0] == '#':
        if (
                all(map(lambda x: x in "?#", xs[:instructions[0]]))
            ) and (
                    xs[instructions[0]] in ".?"
                    ):
                return countConfigs(xs[instructions[0]+1:], instructions[1:])
        return 0
    elif xs[0] == "?":
        return (
                countConfigs(xs[1:], instructions)
                +
                countConfigs("#" + xs[1:], instructions)
                )
result1 = 0
result2 = 0

for i, rawIn in enumerate(sys.stdin):
    splittedInput = rawIn.split(" ")
    xs = splittedInput[0]
    instructions = [int(x) for x in splittedInput[1].split(",")]
    result1 += countConfigs(xs, tuple(instructions))
    result2 += countConfigs((5*("?"+xs))[1:], tuple(5*instructions))

print("Part 1:", result1)
print("Part 2:", result2)
